types/ckeditor__ckeditor5-engine/v27/src/view/matcher.d.ts

```diff
@@ -1,49 +1,62 @@
-import Element from "./element";
+import Element from './element';
 
 export default class Matcher {
-    constructor(
-        pattern:
-            | string
-            | RegExp
-            | {
-                  attributes?: Record<string, string | RegExp | boolean> | undefined;
-                  classes?: string | RegExp | Array<string | RegExp> | undefined;
-                  name?: string | RegExp | undefined;
-                  styles: Record<string, string>;
-              },
-    );
-    add(
-        ...pattern: Array<{
-            attributes?: Record<string, string | RegExp | boolean> | undefined;
-            classes?: string | RegExp | Array<string | RegExp> | undefined;
-            name?: string | RegExp | undefined;
-            styles: Record<string, string>;
-        }>
-    ): void;
+    constructor(...pattern: MatcherPattern[]);
+    add(...pattern: MatcherPattern[]): void;
     getElementName(): string | null;
-    match(
-        element: Element,
-    ): {
+    match(...element: Element[]): {
         element: Element;
-        match: { name?: boolean | undefined; attribute?: string[] | undefined; classes?: string[] | undefined; styles?: Array<[string, string]> | undefined };
+        pattern: MatcherPattern;
+        match: {
+            name?: boolean | undefined;
+            attribute?: string[] | undefined;
+            classes?: string[] | undefined;
+            styles?: Array<[string, string]> | undefined;
+        };
     } | null;
-    matchAll(
-        element: Element,
-    ): Array<{
+    matchAll(...element: Element[]): Array<{
         element: Element;
-        match: { name?: boolean | undefined; attribute?: string[] | undefined; classes?: string[] | undefined; styles?: Array<[string, string]> | undefined };
+        match: {
+            name?: boolean | undefined;
+            attribute?: string[] | undefined;
+            classes?: string[] | undefined;
+            styles?: Array<[string, string]> | undefined;
+        };
     }> | null;
 }
 
 export type MatcherPattern =
-    | ((
-          element: Element,
-      ) => null | { name?: boolean | undefined; attribute?: string[] | undefined; classes?: string[] | undefined; styles?: Array<[string, string]> | undefined })
+    | ((element: Element) => null | {
+          name?: boolean | undefined;
+          attribute?: string[] | undefined;
+          classes?: string[] | undefined;
+          styles?: Array<[string, string]> | undefined;
```

Comment:
I think this is conflating the type of `result.match` returned from `match` and `matchAll` and the types of the patterns passed to add and the constructor. Here's the jsdoc for the latter:

https://github.com/ckeditor/ckeditor5/blob/db4c718490ad4551c17e188be7cd02960486b7b3/packages/ckeditor5-engine/src/view/matcher.js#L55-L70

So it seems that if the matcher pattern is a function, its type is actually `(element: Element) => ReturnType<Matcher.match>["match"]` (as you have it above, i.e. `{name?: boolean; attribute: string[]; ...}` ) rather than `((element: Element) => null`, and the other types in the union are what you had originally in `constructor` and `add` before this commit. So altogether I think it's supposed to be:

```typescript
export type MatcherPattern =
    | ((element: Element) => {
        name?: boolean | undefined;
        attribute?: string[] | undefined;
        classes?: string[] | undefined;
        styles?: Array<[string, string]> | undefined;
      })
    | string
    | RegExp
    | {
          attributes?: Record<string, string | RegExp | boolean> | undefined;
          classes?: string | RegExp | Array<string | RegExp> | undefined;
          name?: string | RegExp | undefined;
          styles?: Record<string, string> | undefined;
      };

Reply:
That totally makes sense. I'll update next week then.

Reply:
@fdintino It seems that the function may return `void` too.

https://github.com/ckeditor/ckeditor5/blob/master/packages/ckeditor5-page-break/src/pagebreakediting.js#L95

Reply:
Ah, good catch.